include/alpaka/queue/sycl/QueueGenericSyclBase.hpp

```diff
@@ -0,0 +1,252 @@
+                m_queue.wait_and_throw();
+            }
+            catch(sycl::exception const& err)
+            {
+                std::cerr << "Caught SYCL exception while destructing a SYCL queue: " << err.what() << std::endl;
+            }
+            catch(std::runtime_error const& err)
+            {
+                std::cerr << "The following runtime errors occured while destructing a SYCL queue:" << std::endl;
+                std::cerr << err.what() << std::endl;
+            }
+        }
+
+        // Don't call this without locking first!
+        auto clean_dependencies() -> void
+        {
+            // Clean up completed events
+            auto const start = std::begin(m_dependencies);
+            auto const old_end = std::end(m_dependencies);
+            auto const new_end = std::remove_if(
+                start,
+                old_end,
+                [](sycl::event ev) {
+                    return ev.get_info<sycl::info::event::command_execution_status>()
+                        == sycl::info::event_command_status::complete;
+                });
+
+            m_dependencies.erase(new_end, old_end);
+        }
+
+        auto register_dependency(sycl::event event) -> void
+        {
+            std::lock_guard<std::shared_mutex> lock{m_mutex};
+
+            clean_dependencies();
+            m_dependencies.push_back(event);
+        }
+
+        auto empty() const -> bool
+        {
+            std::shared_lock<std::shared_mutex> lock{m_mutex};
+            return m_last_event.get_info<sycl::info::event::command_execution_status>()
+                == sycl::info::event_command_status::complete;
+        }
+
+        auto wait() -> void
+        {
+            std::lock_guard<std::shared_mutex> lock{m_mutex};
+            m_queue.wait_and_throw();
+        }
+
+        auto get_last_event() const -> sycl::event
+        {
+            std::shared_lock<std::shared_mutex> lock{m_mutex};
+            return m_last_event;
+        }
+
+        template<bool TBlocking, typename TTask>
+        auto enqueue(TTask const& task) -> void
+        {
+            std::lock_guard<std::shared_mutex> lock{m_mutex};
+
+            clean_dependencies();
+
+            // Execute task
+            m_last_event = m_queue.submit(
+                [this, &task](sycl::handler& cgh)
+                {
+                    if(!m_dependencies.empty())
+                        cgh.depends_on(m_dependencies);
+
+                    if constexpr(is_sycl_kernel<TTask>) // Kernel
+                        task(cgh, m_fence_dummy); // Will call cgh.parallel_for internally
+                    else if constexpr(is_sycl_task<TTask>) // Copy / Fill
+                        task(cgh); // Will call cgh.{copy, fill} internally
+                    else // Host
+                        cgh.host_task(task);
+                });
+
+            m_dependencies.clear();
+
+            if(TBlocking)
+                m_queue.wait_and_throw();
+        }
```

Comment:
```suggestion
        template<bool TBlocking, typename TTask>
        auto enqueue(TTask const& task) -> void
        {
            {
            std::lock_guard<std::shared_mutex> lock{m_mutex};

            clean_dependencies();

            // Execute task
            m_last_event = m_queue.submit(
                [this, &task](sycl::handler& cgh)
                {
                    if(!m_dependencies.empty())
                        cgh.depends_on(m_dependencies);

                    if constexpr(is_sycl_kernel<TTask>) // Kernel
                        task(cgh, m_fence_dummy); // Will call cgh.parallel_for internally
                    else if constexpr(is_sycl_task<TTask>) // Copy / Fill
                        task(cgh); // Will call cgh.{copy, fill} internally
                    else // Host
                        cgh.host_task(task);
                });

            m_dependencies.clear();
            }

            if(TBlocking)
                m_queue.wait_and_throw();
        }
```
I think you can run the `m_queue.wait_and_throw();` outside the locked region.